src/tir/schedule/primitive/rolling_buffer.cc

```diff
@@ -0,0 +1,443 @@
+    if (stmt->buffer.same_as(info_->old_buffer)) {
+      Array<PrimExpr> new_indices;
+      new_indices.reserve(stmt->indices.size());
+      // First modify the access indices to use modulo arithmetic
+      // for the rolling axis
+      for (size_t i = 0; i < stmt->indices.size(); ++i) {
+        auto index = stmt->indices[i];
+        if (static_cast<int>(i) == info_->rolling_axis) {
+          new_indices.push_back(FloorMod(index, info_->rolling_extent));
+        } else {
+          new_indices.push_back(index);
+        }
+      }
+      BufferStoreNode* n = stmt.CopyOnWrite();
+      // Replace the stored buffer with the new buffer.
+      n->buffer = info_->new_buffer;
+      n->indices = std::move(new_indices);
+      // Need to add predicate to the current block to avoid recomputing elements.
+      rewrite_block_predicate_ = true;
+    }
+    return std::move(stmt);
+  }
+
+  PrimExpr VisitExpr_(const BufferLoadNode* op) final {
+    BufferLoad stmt = Downcast<BufferLoad>(StmtExprMutator::VisitExpr_(op));
+    if (stmt->buffer.same_as(info_->old_buffer)) {
+      Array<PrimExpr> new_indices;
+      new_indices.reserve(stmt->indices.size());
+      for (size_t i{0}; i < stmt->indices.size(); ++i) {
+        auto index = stmt->indices[i];
+        if (static_cast<int>(i) == info_->rolling_axis) {
+          new_indices.push_back(FloorMod(index, info_->rolling_extent));
+        } else {
+          new_indices.push_back(index);
+        }
+      }
+      BufferLoadNode* n = stmt.CopyOnWrite();
+      // Replace the loaded buffer with the new buffer.
+      n->buffer = info_->new_buffer;
+      n->indices = std::move(new_indices);
+    }
+    return std::move(stmt);
+  }
+
+ private:
+  const StmtSRef& scope_sref_;
+  RollingBufferInfo* info_;
+  bool rewrite_block_predicate_ = false;
+};
+
+}  // namespace
+
+void RollingBuffer(ScheduleState self, const StmtSRef& block_sref, int write_buffer_index) {
+  /*!
+   *  Check
+   *    - The block is not an output block.
+   *    - The block is tiled and there is access overlap between adjacent tiles.
+   *  Mutate
+   *    - Select the outermost rollable axis appeared in the block's loop nest
+   *      as the 'rolling axis', trim the target buffer from the rolling axis.
+   *    - Use modulo arithmetic to modify the target buffer's read and load
+   *      indices to circularize the buffer along the rolling dimension.
+   *    - Append block predicate to avoid recomputing overlapping elements.
+   */
+  Map<Var, arith::IntSet> dom_map;
+  const BlockRealize& realize = GetBlockRealize(self, block_sref);
+  const Block& block = realize->block;
+
+  // Step 1. Checking index, getting the target buffer region and the parent scope.
+  const BufferRegion& buffer_region =
+      GetNthAccessBufferRegion(self, block, write_buffer_index, BufferIndexType::kWrite);
+  StmtSRef scope_root_sref = GetScopeRoot(self, block_sref, /*require_stage_pipeline=*/false);
```

Comment:
The `rolling_buffer ` primitive itself leads to `stage_pipeline`  of the root block to be false, because the number of elements calculated per tile of the rolling buffer does not contain overlapping elements. If we set `require_stage_pipeline `as true, `rolling_buffer` can no longer be applied to other blocks under the root block, the test case  `test_cascade_3_max_pool2d_with_stride` would fail.
So I suggest to set `require_stage_pipeline` as false here to adapt to the situation where there are more than two cascading layers.

Reply:
Ok, that makes sense to me. I agree use `require_stage_pipeline=False`. Then could we try best to check there is no complex rw dependencies. Eg, the consumer-producer relation is unique?